### PR TITLE
UnionSet

### PR DESCRIPTION
**Pull request by mforets**
See .https://github.com/JuliaReach/LazySets.jl/issues/824.

Note that `UnionSet` was removed previously, see [this PR](https://github.com/JuliaReach/LazySets.jl/pull/266).

### CODE REVIEW

File: src/UnionSet.jl [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+- `array` -- array of convex sets
+"""
+struct UnionSetArray{N<:Real, S<:LazySet{N}}
+    array::Vector{S}
+
+    # default constructor with dimension check
+    function UnionSetArray{N, S}(X::Vector{S}) where {N<:Real, S<:LazySet{N}}
+        @assert !isempty(X) "the array in this `UnionSetArray` is empty"
+        n = dim(X[1])
+        @assert all(x -> dim(x) == n, X) "sets in a union must have the same dimension"
```
Comment (schillic) [inline src/UnionSet.jl]: Hm, we do not have such a thing for other array types. I would rather do this in #759.

Comment (mforets) [reply src/UnionSet.jl]: Alright, i'm removing that inner constructor.

**Comment by mforets** [comment]
I've declared `UnionSetArray` as the array version of `UnionSet` with the `@declare_array_version` macro. Seems like there is no need to make `UnionSet` inherit from `LazySet` to use it.

File: src/UnionSet.jl [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+Return the support vector of the union of two convex sets in a given direction.
+
+### Input
+
+- `d`         -- direction
+- `cap`       -- union of two convex sets
+- `algorithm` -- (optional, default: "support_vector"): the algorithm to compute
+                 the support vector; if "support_vector", use the support
+                 vector of each argument; if "support_function" use the support
+                 function of each argument and evalute the support vector of only
```
Comment (schillic) [inline src/UnionSet.jl]:
```suggestion
                 function of each argument and evaluate the support vector of only
```

File: src/UnionSet.jl [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+a given direction.
+
+### Input
+
+- `d`         -- direction
+- `cap`       -- union of a finite number of convex sets
+- `algorithm` -- (optional, default: "support_vector"): the algorithm to compute
+                 the support vector; if "support_vector", use the support
+                 vector of each argument; if "support_function" use the support
+                 function of each argument and evalute the support vector of only
```
Comment (schillic) [inline src/UnionSet.jl]:
```suggestion
                 function of each argument and evaluate the support vector of only
```

File: src/UnionSet.jl [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+### Output
+
+The array that holds the union of a finite number of convex sets.
+"""
+function array(cap::UnionSetArray{N, S})::Vector{S} where {N<:Real, S<:LazySet{N}}
+    return cap.array
+end
+
+"""
+    σ(d::AbstractVector{N}, cap::UnionSetArray{N}; algorithm="support_vector") where {N<:Real}
```
Comment (schillic) [inline src/UnionSet.jl]:
```suggestion
    σ(d::AbstractVector{N}, cap::UnionSetArray{N}; [algorithm]="support_vector") where {N<:Real}
```

File: src/UnionSet.jl [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+### Output
+
+The ambient dimension of the union of two convex sets.
+"""
+function dim(cup::Union)::Int
+    return dim(cup.X)
+end
+
+"""
+    σ(d::AbstractVector{N}, cap::UnionSet{N}; algorithm="support_vector") where {N<:Real}
```
Comment (schillic) [inline src/UnionSet.jl]:
```suggestion
    σ(d::AbstractVector{N}, cap::UnionSet{N}; [algorithm]="support_vector") where {N<:Real}
```

File: src/UnionSet.jl [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+can be obtained as the vector that maximizes the support function, i.e.
+it is sufficient to find the ``\\argmax(ρ(d, X₂), ρ(d, X₂), ...])`` and evaluate
+its support vector.
+
+The default implementation, with option `algorithm="support_vector"`, computes
+the support vector of all ``X₁, X₂, ...`` and then compares the support function using
+a dot product. If it happens that the support function can be more efficiently
+computed (without passing through the support vector), consider using the alternative
+`algorithm="support_function"` implementation, which evaluates the support function
+of each set directly and then calls only the support vector of only one of the ``Xᵢ``.
```
Comment (schillic) [inline src/UnionSet.jl]:
```suggestion
of each set directly and then calls only the support vector of one of the ``Xᵢ``.
```

File: src/UnionSet.jl (hunk 5 above)
Comment (schillic) [inline src/UnionSet.jl]: `cap` → `cup` (all over this file)

File: src/UnionSet.jl [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+"""
+function σ(d::AbstractVector{N}, cap::UnionSetArray{N}; algorithm="support_vector") where {N<:Real}
```
Comment (schillic) [inline src/UnionSet.jl]: These algorithms can be implemented in constant memory, but let us keep yours for simplicity.

Comment (mforets) [reply src/UnionSet.jl]: Do open an issue if you have an idea for improvement of this one, to modify `"support_function`, `"support_vector"` or add a new one.

Comment (schillic) [reply src/UnionSet.jl]: See #993.

**Pull request merged**